### PR TITLE
subscriber: add `MakeWriter` combinators

### PR DESCRIPTION
**Pull request by hawkw**
Depends on #1141.

This branch adds a `MakeWriterExt` trait which adds a number of
combinators for composing multiple types implementing `MakeWriter`.
`MakeWriter`s can now be teed together, filtered with minimum and
maximum levels, filtered with a `Metadata` predicate, and combined with
a fallback for when a writer is _not_ made for a particular `Metadata`.

I've also added a `MakeWriter` impl for `std::fs::File`, because `&File`
implements `Write`. Ideally, we'd have a default impl of `MakeWriter`
for `T where &T: Write`, but that's not possible due to the impl of
`MakeWriter` for `F: Fn() -> T: Write`. We could add a generic `by_ref`
adapter for any `T where &T: Write`, though...

### CODE REVIEW

**Review by davidbarsky — CHANGES_REQUESTED** [review]
This looks _super_ cool, modulo warnings and docs! Mind re-upping me for review once that's done?

File: tracing-subscriber/src/fmt/writer.rs [hunk 1]
```diff
@@ -223,6 +223,253 @@ pub trait MakeWriter<'a> {
...
+    fn with_min_level(self, level: tracing_core::Level) -> WithMinLevel<Self>
+    where
+        Self: Sized,
+    {
+        WithMinLevel::new(self, level)
+    }
+
+    /// Wraps `self` with a predicate that takes a span or event's [`Metadata`]
+    /// and returns a `bool`. The returned [`MakeWriter`]'s
+    /// [`MakeWriter::make_writer_for`][mwf] method will check the predicate to
```
[the remark this answers is not in the document]
Comment (hawkw) [reply tracing-subscriber/src/fmt/writer.rs:341]: whoops, my bad --- i forgot to add that.

File: tracing-subscriber/src/fmt/writer.rs [hunk 2]
```diff
@@ -223,6 +223,253 @@ pub trait MakeWriter<'a> {
...
 
+/// Extension trait adding combinators for working with types implementing
+/// [`MakeWriter`].
+///
+/// This is not intended to be implemented directly for user-defined
+/// [`MakeWriter`]s; instead, it should be imported when the desired methods are
+/// used.
+pub trait MakeWriterExt<'a>: MakeWriter<'a> {
+    /// Wraps `self` and returns a [`MakeWriter`] that will only write output
+    /// for events at or below the provided verbosity [`Level`].
```
[the remark this answers is not in the document]
Comment (hawkw) [reply tracing-subscriber/src/fmt/writer.rs:234]: hmm, the docs for `Level` mention a concept of "verbosity" but don't actually describe the comparison behavior: https://docs.rs/tracing/0.1.26/tracing/struct.Level.html we should probably add something there as well, in a separate branch. I'll merge this change for now though.

[the remark this answers is not in the document]
Comment (hawkw) [reply tracing-subscriber/src/fmt/writer.rs]: opened #1446 to add a (fairly excessive) docs section explaining levels and level comparisons

**Review by davidbarsky — APPROVED** [review]

**Pull request merged**